auth/auth.go

```diff
@@ -222,20 +218,51 @@ func (auth *Authenticator) calculateHistory(invalSeq uint64, invalGrants ch.Time
 			currentHistoryForGrant = GrantHistory{}
 		}
 
-		// TODO: Will perform pruning here once full
-
 		// Add grant to history
-		currentHistoryForGrant.UpdatedAt = time.Now().UnixNano()
+		currentHistoryForGrant.UpdatedAt = time.Now().Unix()
 		currentHistoryForGrant.Entries = append(currentHistoryForGrant.Entries, GrantHistorySequencePair{
 			StartSeq: previousInfo.Sequence,
 			EndSeq:   invalSeq,
 		})
 		currentHistory[previousName] = currentHistoryForGrant
 	}
 
+	if prunedHistory := currentHistory.PruneHistory(auth.ClientPartitionWindow); len(prunedHistory) > 0 {
+		base.Debugf(base.KeyCRUD, "rebuildChannels: Pruned principal history on %s for %s", base.UD(princName), base.UD(prunedHistory))
+	}
+
+	// Ensure no entries are larger than the allowed threshold
+	maxHistoryEntriesPerGrant := CalculateMaxHistoryEntriesPerGrant(len(currentHistory))
+	for grantName, grantHistory := range currentHistory {
+		if len(grantHistory.Entries) > maxHistoryEntriesPerGrant {
+			grantHistory.Entries[1].StartSeq = grantHistory.Entries[0].StartSeq
+			currentHistory[grantName] = grantHistory
+		}
+	}
+
 	return currentHistory
 }
 
+func CalculateMaxHistoryEntriesPerGrant(channelCount int) int {
+	const maximumHistoryBytes = 1024 * 1024 // 1MB
```

Comment:
All three of these constants are the same units, but that's not evident from naming.  What do you think about maximumHistoryBytes, averageHistoryKeyBytes, and averageHistoryEntryBytes?